modules/core/src/test/java/org/apache/ignite/internal/encryption/AbstractEncryptionTest.java

```diff
@@ -271,4 +313,185 @@ protected boolean checkMasterKeyName(String name) {
+                continue;
+
+            GridEncryptionManager encryption = grid.context().encryption();
+
+            assertEquals(grid.localNode().id().toString(), (byte)keyId, encryption.groupKey(grpId).id());
+
+            IgniteInternalFuture<Void> fut = encryption.reencryptionFuture(grpId);
+
+            // The future will be completed after the checkpoint, forcecheckpoint does nothing
+            // if the checkpoint has already been scheduled.
+            GridTestUtils.waitForCondition(() -> {
+                if (fut.isDone())
+                    return true;
+
+                try {
+                    forceCheckpoint(g);
+                }
+                catch (IgniteCheckedException e) {
+                    throw new RuntimeException(e);
+                }
+
+                return fut.isDone();
+            }, timeout);
+
+            assertTrue(fut.isDone());
+
+            CacheGroupContext grp = grid.context().cache().cacheGroup(grpId);
+
+            List<Integer> parts = IntStream.range(0, grp.shared().affinity().affinity(grpId).partitions())
+                .boxed().collect(Collectors.toList());
+
+            parts.add(INDEX_PARTITION);
+
+            int realPageSize = grp.dataRegion().pageMemory().realPageSize(grpId);
+            int encryptionBlockSize = grp.shared().kernalContext().config().getEncryptionSpi().blockSize();
+
+            for (int p : parts) {
+                FilePageStore pageStore =
+                    (FilePageStore)((FilePageStoreManager)grp.shared().pageStore()).getStore(grpId, p);
+
+                if (!pageStore.exists())
+                    continue;
+
+                long state = grid.context().encryption().getEncryptionState(grpId, p);
+
+                String msg = String.format("p=%d, off=%d, total=%d",
+                    p, ReencryptStateUtils.pageIndex(state), ReencryptStateUtils.pageCount(state));
+
+                assertEquals(msg, 0, ReencryptStateUtils.pageCount(state));
+                assertEquals(msg, 0, ReencryptStateUtils.pageIndex(state));
+
+                long startPageId = PageIdUtils.pageId(p, PageIdAllocator.FLAG_DATA, 0);
+
+                int pagesCnt = pageStore.pages();
+                int pageSize = pageStore.getPageSize();
+
+                ByteBuffer pageBuf = ByteBuffer.allocate(pageSize);
+
+                Path path = new File(pageStore.getFileAbsolutePath()).toPath();
+
+                try (FileChannel ch = FileChannel.open(path, StandardOpenOption.READ)) {
+                    for (int n = 0; n < pagesCnt; n++) {
+                        long pageId = startPageId + n;
+                        long pageOff = pageStore.pageOffset(pageId);
+
+                        pageBuf.position(0);
+
+                        ch.position(pageOff);
+                        ch.read(pageBuf);
+
+                        pageBuf.position(realPageSize + encryptionBlockSize);
+
+                        // If crc present
```

Comment:
Point at the end.
Why there can be pages without CRC?

Reply:
CRC can be calculated as "0" - it was my mistake.
But in rare cases, when scanning a partition on a disk, we find a blank page. I added a special check for this case.